### DISCUSSION TITLE
explore more efficient approaches to inscription

### DISCUSSION THREAD

**Discussion by conduition** [ordinals/ord#2879] [Ideas]
Hey ordinal devs. There's [a lot of debate going on in bitcoin core](https://github.com/bitcoin/bitcoin/pull/28408) about whether to label inscription transactions as non-standard, thus preventing their propagation through the mempool. 

My personal perspective is summed up in [this article](https://conduition.io/bitcoin/inviscriptions/) where I demonstrate that preventing or even identifying arbitrary data upload (i.e. inscribing) is actually impossible.

I'm hoping that this issue can serve as a concrete discussion of how to lessen the footprint of inscriptions on the bitcoin mempool, ideally in a way that meaningfully improves fee market conditions on the Bitcoin mainnet.

Some possible paths to explore:
- Analyze which types of inscriptions are taking up the most blockchain/mempool space today. This will tell us where effort is best spent.
- Dedicate specific magic byte markers for common content types. E.g. substitute `\x12\x34` (2 bytes) in place of the content type header `image/jpeg` (10 bytes). 
- (BRC-20) Use binary encoding. An inscription like [this one](https://mempool.space/tx/2729c4a16477592e1a0ff73f3ee60aa75436663c5bfda085f53c87f615504390) [in text/utf8: `{"p":"brc-20","op":"mint","tick":"iost","amt":"1000"}`] could easily be compressed into just 5 or 6 bytes, instead of taking up 53 bytes of witness space. This would need to be done upstream in https://github.com/unisat-wallet/libbrc20-indexer and [the spec](https://layer1.gitbook.io/layer1-foundation/foundation/structure)
- (NFTs) Instead of uploading large blobs of data to the blockchain, store the blob on some other decentralized storage system, such as a torrent, IPFS hash, or Nostr note. Then commit only a reference to the blockchain.
- If the user insists on pushing data to the chain, use a compression scheme by default. [This NFT inscription, for example](https://ordiscan.com/inscription/34595802), could have saved 4 kilobytes of witness space, 1000 vbytes, or enough for 5+ normal bitcoin transactions, just by wrapping its inner webp image in XZIP/GZIP.

**Comment by conduition** [comment]
Related: https://l1f.discourse.group/t/brc-20-optimization/33

**Reply by conduition** [reply]
The savings from content-type aliases would be nice, but they're nothing compared to what could be saved by optimizing the BRC20 inscription body format from JSON into a concise binary format. @casey if you know anyone who could help me action this, I'd be very grateful :pray:

**Comment by conduition** [comment]
> - Analyze which types of inscriptions are taking up the most blockchain/mempool space today. This will tell us where effort is best spent.

I'm currently waiting on `ord` to finish indexing so i can run some analysis and get some hard numbers on which parts and kinds of inscriptions are impacting the blockchain most heavily.

**Reply by conduition** [reply]
Is it normal for this process to take 30 to 60 seconds to parse a single block? (built with `--release`)

**Reply by conduition** [reply]
Some results posted here: https://github.com/ordinals/ord/discussions/2879#discussioncomment-7930615

**Comment by conduition** [comment]
@casey @raphjaph I think this one is a no-brainer.  

> - Dedicate specific magic byte markers for common content types. E.g. substitute `\x12\x34` (2 bytes) in place of the content type header `image/jpeg` (10 bytes). 

Thoughts?

**Reply by casey** [reply]
I think this might have limited impact. If we replaced every content type string with a single byte, by what percentage would the total size of inscriptions be reduced by?

**Reply by conduition** [reply]
A BRC-20 inscription envelope is about 70-80 bytes, with a content type of `text/plain` (length 10). Subtracting 9 bytes, that saves more than 11% off the total inscription size.

**Reply by conduition** [reply]
This data is going onto an immutable distributed ledger. Every byte counts

**Reply by raphjaph** [reply]
The nice thing about using standard MIME types is that people and clients can inscribe and show any kind of content and do not have to translate between magic bytes. You probably save a little but not enough to justify the inconvenience of deciding and translating every kind of MIME type into a magic byte.

**Reply by conduition** [reply]
> You probably save a little but not enough to justify the inconvenience of deciding and translating every kind of MIME type into a magic byte.

You wouldn't need to assign magic bytes to _every_ content type, just the most common ones. If someone wants to use a wacky edgecase MIME type, they can include the full MIME-type string, but they would lose out on efficiency from using more concise encoding. 

What's so "inconvenient" about an optional encoding layer between magic bytes and MIME types? Are you talking about DX?

**Reply by casey** [reply]
Another thing regarding BRC-20, they actually don't need the content type, since that's just for display, so the most efficient thing they could do would be to omit the content type entirely, and try to parse every envelope.

**Reply by conduition** [reply]
So I gave up on waiting for `ord` to sync and wrote my own simple indexer which doesn't track sat ownership, just inscription metrics. 

I indexed the last 100 blocks (822333 - 822433) and analyzed each of the inscriptions contained in those blocks.

If every content type header had been replaced with a single byte, then **each block could have saved on average 35,932 witness bytes.** That's 8983 vbytes **per block** being wasted, or enough for about 45 normal transactions.



<details>
<summary><h3>per block results</h3></summary>

```
822333: 39665 content type bytes; 37818 bytes could have been saved
822334: 50854 content type bytes; 48431 bytes could have been saved
822335: 17149 content type bytes; 16165 bytes could have been saved
822336: 51994 content type bytes; 49684 bytes could have been saved
822337: 82519 content type bytes; 78901 bytes could have been saved
822338: 87951 content type bytes; 83992 bytes could have been saved
822339: 40773 content type bytes; 38840 bytes could have been saved
822340: 67673 content type bytes; 64345 bytes could have been saved
822341: 79386 content type bytes; 75634 bytes could have been saved
822342: 50019 content type bytes; 46763 bytes could have been saved
822343: 43199 content type bytes; 40828 bytes could have been saved
822344: 61382 content type bytes; 58429 bytes could have been saved
822345: 48252 content type bytes; 45887 bytes could have been saved
822346: 36214 content type bytes; 34569 bytes could have been saved
822347: 54560 content type bytes; 52127 bytes could have been saved
822348: 41579 content type bytes; 39543 bytes could have been saved
822349: 56300 content type bytes; 53246 bytes could have been saved
822350: 68593 content type bytes; 65528 bytes could have been saved
822351: 49337 content type bytes; 46592 bytes could have been saved
822352: 64291 content type bytes; 61212 bytes could have been saved
822353: 45528 content type bytes; 43372 bytes could have been saved
822354: 52545 content type bytes; 50250 bytes could have been saved
822355: 58069 content type bytes; 54690 bytes could have been saved
822356: 45174 content type bytes; 42794 bytes could have been saved
822357: 21867 content type bytes; 20815 bytes could have been saved
822358: 47278 content type bytes; 44017 bytes could have been saved
822359: 121576 content type bytes; 116361 bytes could have been saved
822360: 31482 content type bytes; 29893 bytes could have been saved
822361: 49469 content type bytes; 47276 bytes could have been saved
822362: 27752 content type bytes; 26397 bytes could have been saved
822363: 64959 content type bytes; 62045 bytes could have been saved
822364: 154000 content type bytes; 147581 bytes could have been saved
822365: 101485 content type bytes; 97209 bytes could have been saved
822366: 120436 content type bytes; 115391 bytes could have been saved
822367: 10688 content type bytes; 10227 bytes could have been saved
822368: 22398 content type bytes; 21270 bytes could have been saved
822369: 8332 content type bytes; 7941 bytes could have been saved
822370: 3318 content type bytes; 3176 bytes could have been saved
822371: 15397 content type bytes; 14696 bytes could have been saved
822372: 19925 content type bytes; 19018 bytes could have been saved
822373: 17073 content type bytes; 16307 bytes could have been saved
822374: 14833 content type bytes; 14197 bytes could have been saved
822375: 28434 content type bytes; 27148 bytes could have been saved
822376: 23572 content type bytes; 22439 bytes could have been saved
822377: 31124 content type bytes; 29660 bytes could have been saved
822378: 19635 content type bytes; 18721 bytes could have been saved
822379: 10891 content type bytes; 10318 bytes could have been saved
822380: 6103 content type bytes; 5790 bytes could have been saved
822381: 24263 content type bytes; 23186 bytes could have been saved
822382: 18786 content type bytes; 17858 bytes could have been saved
822383: 21358 content type bytes; 20304 bytes could have been saved
822384: 14110 content type bytes; 13377 bytes could have been saved
822385: 9008 content type bytes; 8590 bytes could have been saved
822386: 23424 content type bytes; 22274 bytes could have been saved
822387: 37691 content type bytes; 35917 bytes could have been saved
822388: 5271 content type bytes; 5015 bytes could have been saved
822389: 23344 content type bytes; 22054 bytes could have been saved
822390: 11537 content type bytes; 10982 bytes could have been saved
822391: 39243 content type bytes; 37410 bytes could have been saved
822392: 18619 content type bytes; 17629 bytes could have been saved
822393: 13997 content type bytes; 13238 bytes could have been saved
822394: 29251 content type bytes; 27759 bytes could have been saved
822395: 32797 content type bytes; 31346 bytes could have been saved
822396: 66536 content type bytes; 63647 bytes could have been saved
822397: 3357 content type bytes; 3199 bytes could have been saved
822398: 16262 content type bytes; 15522 bytes could have been saved
822399: 45600 content type bytes; 43631 bytes could have been saved
822400: 8416 content type bytes; 7990 bytes could have been saved
822401: 16238 content type bytes; 15502 bytes could have been saved
822402: 24085 content type bytes; 22914 bytes could have been saved
822403: 24072 content type bytes; 22984 bytes could have been saved
822404: 28903 content type bytes; 27556 bytes could have been saved
822405: 29438 content type bytes; 27964 bytes could have been saved
822406: 32053 content type bytes; 30508 bytes could have been saved
822407: 18627 content type bytes; 17794 bytes could have been saved
822408: 13443 content type bytes; 12848 bytes could have been saved
822409: 23594 content type bytes; 22462 bytes could have been saved
822410: 47871 content type bytes; 45656 bytes could have been saved
822411: 23041 content type bytes; 21808 bytes could have been saved
822412: 8327 content type bytes; 7784 bytes could have been saved
822413: 13818 content type bytes; 13032 bytes could have been saved
822414: 46067 content type bytes; 43696 bytes could have been saved
822415: 127198 content type bytes; 121628 bytes could have been saved
822416: 79918 content type bytes; 76482 bytes could have been saved
822417: 100672 content type bytes; 96230 bytes could have been saved
822418: 58095 content type bytes; 55271 bytes could have been saved
822419: 22086 content type bytes; 21025 bytes could have been saved
822420: 6587 content type bytes; 6295 bytes could have been saved
822421: 9200 content type bytes; 8780 bytes could have been saved
822422: 8509 content type bytes; 8087 bytes could have been saved
822423: 10337 content type bytes; 9843 bytes could have been saved
822424: 14322 content type bytes; 13636 bytes could have been saved
822425: 18601 content type bytes; 17640 bytes could have been saved
822426: 15319 content type bytes; 14631 bytes could have been saved
822427: 45514 content type bytes; 43305 bytes could have been saved
822428: 75262 content type bytes; 71966 bytes could have been saved
822429: 26475 content type bytes; 25211 bytes could have been saved
822430: 11690 content type bytes; 11149 bytes could have been saved
822431: 22369 content type bytes; 21277 bytes could have been saved
822432: 28469 content type bytes; 27079 bytes could have been saved
822433: 13426 content type bytes; 12772 bytes could have been saved
```

</details>


@raphjaph @casey That's significant. And it can easily be fixed. You wouldn't even need to introduce a huge map; just hitting the top 10 mime types like `text/plain` would do the trick. Here's the most commonly used content types in the last 100 blocks sorted by total witness byte consumption. 

```
text/plain;charset=utf-8: 3292152
text/plain: 325620
text/html;charset=utf-8: 75877
application/json;charset=utf-8: 32070
image/png: 21798
model/gltf-binary: 16048
image/webp: 5850
application/json: 912
image/jpeg: 510
image/svg+xml: 273
text/html: 81
DATA: 60
text/plain; charset=utf-8: 50
...misc omitted
```

If it would be useful, i can also get you data on earlier blocks. As you can see from [this metrics dashboard](https://dune.com/dataalways/ordinals), BRC20 didn't always take up this much space.

**Reply by casey** [reply]
I think I'm warming up to this a little bit. One downside is that it would be an ecosystem wide change, if ord started doing it, other indexers would need to upgrade too.

**Reply by casey** [reply]
And thanks for the detailed data!

**Reply by conduition** [reply]
With pleasure :smile: 

> One downside is that it would be an ecosystem wide change, if ord started doing it, other indexers would need to upgrade too.

Yes, updated indexers would be backwards-compatible, but old indexers would not recognize the new content type 'aliases'.

To avoid causing UTF8 decoding errors in old clients, you could use valid ASCII characters as the aliases, since you don't need many of them.

e.g:

| Byte | ASCII Character | Content-Type Meaning | 
|:-:|:-:|:-:|
|0x74| `t` | `text/plain;charset=utf-8` |
|0x6A| `j` | `application/json` |
|0x68| `h` | `text/html` |

This way, older clients would still recognize inscriptions which use the new aliases, although they would appear to have weird one-character content-types.

**Reply by casey** [reply]
I'm not sure overloading the same field is a good idea, it might be better to go with a new field.

**Reply by conduition** [reply]
Sure, then older indexers would just see newer inscriptions as having no content type. That's still a valid inscription though IIRC.

I could prepare a PR for this into `ord`, the basics seem pretty straightforward. Do i need to update any docs first so we can agree on a format? Or should the code come first?

**Reply by casey** [reply]
I'm still not 100% sure that the juice is worth the squeeze here. Maybe open an issue, and we can let it marinate for a little bit?

**Reply by conduition** [reply]
https://github.com/ordinals/ord/issues/2944
